### PR TITLE
Addition of missing openstack model / tasks

### PR DESCRIPTION
**Pull request by drekle**
https://github.com/kubernetes/kops/issues/6310
https://github.com/kubernetes/kops/pull/6345

- Added Load Balancer which adds the floating ip to the ca-cert in the same way AWS.
- Changes swift to support environment variables, similar to AWS.
- Designate additions (untested by myself).
- Security Group additions, thought not tightly constrained around the internal network.
- cloud_config additions
- Added feature flag support for the apply cluster flow for the open stack cloud.
I have been testing using the changes which have been in a conversation here:
drekle#1

I plan on raising 3 PR's which encompass all of the changes I have in a develop branch
https://github.com/drekle/kops/tree/develop
where each of the other two branches require changes requested here.  These include a protokube implementation as well as one additional argument required for create_cluster to specify the openstack volume type to be used for the etcd master volumes.

I have been testing with these several branches merged together.  Which is in the full PR here:
https://github.com/kubernetes/kops/pull/6351

One of these PR's should be closed.
```
export KOPS_FEATURE_FLAGS=AlphaAllowOpenstack
export PROTOKUBE_IMAGE=http://<PRIVATE_IP>/.build/upload/kops/1.10.0/images/protokube.tar.gz
export KOPS_BASE_URL=http://<PRIVATE_IP>/.build/upload/kops/1.10.0/
kops create cluster \
  --cloud openstack \
  --name sre.k8s.local \
  --state swift://sre \
  --zones zone-a,zone-b,zone-c \
  --network-cidr 10.1.0.0/16 \
  --image centos \
  --master-count=3 \
  --node-count=1 \
  --node-size 2vCPUx8GB \
  --master-size 2vCPUx8GB \
  --etcd-storage-type CBS \
  --api-loadbalancer-type public \
  --topology private \
  --bastion \
  --networking weave \
  --yes
```
```
kubectl get nodes
NAME                                   STATUS    ROLES     AGE       VERSION
sre.k8s.local-master-zone-a-1          Ready     master    3h        v1.10.12
sre.k8s.local-master-zone-b-1          Ready     master    3h        v1.10.12
sre.k8s.local-master-zone-c-1          Ready     master    3h        v1.10.12
sre.k8s.local-nodes-1                  Ready     node      3h        v1.10.12
```

### CODE REVIEW

**Comment by zetaab** [comment]
Example command returns

unknown flag: --etcd-storage-type

**Comment by zetaab** [comment]
```
./kops create cluster \
  --cloud openstack \
  --name sre2.k8s.local \
  --state ${KOPS_STATE_STORE} \
  --zones zone-1,zone-2,zone-3 \
  --network-cidr 10.1.0.0/16 \
  --image centos \
  --master-count=3 \
  --node-count=1 \
  --node-size m1.medium \
  --master-size m1.medium \
  --api-loadbalancer-type public \
  --topology private \
  --bastion \
  --networking weave
```

```
W0116 11:04:31.430049   57862 executor.go:130] error running task "Subnet/utility-zone-1" (9m58s remaining to succeed): Field cannot be changed: Network
W0116 11:04:31.430082   57862 executor.go:130] error running task "Subnet/utility-zone-3" (9m58s remaining to succeed): Field cannot be changed: Network
W0116 11:04:31.430095   57862 executor.go:130] error running task "Subnet/zone-1" (9m58s remaining to succeed): Field cannot be changed: Network
W0116 11:04:31.430109   57862 executor.go:130] error running task "Subnet/zone-3" (9m58s remaining to succeed): Field cannot be changed: Network
W0116 11:04:31.430119   57862 executor.go:130] error running task "Subnet/utility-zone-2" (9m58s remaining to succeed): Field cannot be changed: Network
W0116 11:04:31.430129   57862 executor.go:130] error running task "Subnet/zone-2" (9m58s remaining to succeed): Field cannot be changed: Network
```

**Comment by zetaab** [comment]
Some comments:

current code does not support multiple kops installations in same openstack project. This is because at least subnets are created without clustername prefix.
```
I0116 11:14:18.206804   59029 subnets.go:184] Assigned CIDR 10.1.32.0/19 to subnet zone-1
I0116 11:14:18.206843   59029 subnets.go:184] Assigned CIDR 10.1.64.0/19 to subnet zone-2
I0116 11:14:18.206859   59029 subnets.go:184] Assigned CIDR 10.1.96.0/19 to subnet zone-3
I0116 11:14:18.206900   59029 subnets.go:198] Assigned CIDR 10.1.0.0/22 to subnet utility-zone-1
I0116 11:14:18.206925   59029 subnets.go:198] Assigned CIDR 10.1.4.0/22 to subnet utility-zone-2
I0116 11:14:18.206939   59029 subnets.go:198] Assigned CIDR 10.1.8.0/22 to subnet utility-zone-3
```

Etcd (master) volume installation style does not work us. We have AvailabilityZone `nova` and then Volumetypes `zone-1` `zone-2` and `zone-3`

Now this installer assumes that VolumeType is always same, and is using az "dynamically" 

```
  Volume/1.etcd-main.sre2.k8s.local
  	AvailabilityZone    	zone-1
  	VolumeType          	zone-1
  Volume/2.etcd-main.sre2.k8s.local
  	AvailabilityZone    	zone-2
  	VolumeType          	zone-1
  Volume/3.etcd-main.sre2.k8s.local
  	AvailabilityZone    	zone-3
  	VolumeType          	zone-1
```

**Comment by zetaab** [comment]
```
./kops create cluster \
  --cloud openstack \
  --name sre.k8s.local \
  --state ${KOPS_STATE_STORE} \
  --zones zone-1,zone-2,zone-3 \
  --network-cidr 10.1.0.0/16 \
  --image centos7-cloudimage-1809 \
  --master-count=3 \
  --node-count=1 \
  --node-size m1.medium \
  --master-size m1.medium \
  --api-loadbalancer-type public \
  --topology private \
  --bastion \
  --networking weave

....

% ./kops update cluster --name sre.k8s.local --yes
I0116 13:10:50.501052   72762 s3context.go:87] Found S3_ENDPOINT="https://s3.ceph.elisa.fi", using as non-AWS S3 backend
W0116 13:10:53.209609   72762 urls.go:71] Using base url from KOPS_BASE_URL env var: "https://kopstest.s3.ceph.elisa.fi/.build/dist/"
W0116 13:10:53.295399   72762 urls.go:163] Using protokube location from PROTOKUBE_IMAGE env var: "https://kopstest.s3.ceph.elisa.fi/.build/dist/images/protokube.tar.gz"
I0116 13:10:54.312860   72762 apply_cluster.go:558] Gossip DNS: skipping DNS validation
I0116 13:10:54.635992   72762 executor.go:103] Tasks: 0 done / 87 total; 41 can run
I0116 13:10:55.471083   72762 vfs_castore.go:729] Issuing new certificate: "ca"
I0116 13:10:55.642713   72762 vfs_castore.go:729] Issuing new certificate: "apiserver-aggregator-ca"
I0116 13:11:03.206609   72762 executor.go:103] Tasks: 41 done / 87 total; 21 can run
I0116 13:11:03.608189   72762 vfs_castore.go:729] Issuing new certificate: "kubelet"
I0116 13:11:03.624370   72762 vfs_castore.go:729] Issuing new certificate: "kube-controller-manager"
I0116 13:11:03.939383   72762 vfs_castore.go:729] Issuing new certificate: "kube-scheduler"
I0116 13:11:04.001392   72762 vfs_castore.go:729] Issuing new certificate: "apiserver-proxy-client"
I0116 13:11:04.127532   72762 vfs_castore.go:729] Issuing new certificate: "kube-proxy"
I0116 13:11:04.275336   72762 vfs_castore.go:729] Issuing new certificate: "kubelet-api"
I0116 13:11:04.372353   72762 vfs_castore.go:729] Issuing new certificate: "kubecfg"
I0116 13:11:04.387198   72762 vfs_castore.go:729] Issuing new certificate: "apiserver-aggregator"
I0116 13:11:04.551799   72762 vfs_castore.go:729] Issuing new certificate: "kops"
I0116 13:11:07.314551   72762 executor.go:103] Tasks: 62 done / 87 total; 11 can run
I0116 13:11:31.336027   72762 executor.go:103] Tasks: 73 done / 87 total; 5 can run
I0116 13:11:35.009746   72762 executor.go:103] Tasks: 78 done / 87 total; 1 can run
I0116 13:11:37.707458   72762 executor.go:103] Tasks: 79 done / 87 total; 4 can run
I0116 13:11:43.074966   72762 context.go:231] hit maximum retries 4 with error CreateFloatingIP: create floating IP failed: Resource not found
W0116 13:11:49.425907   72762 executor.go:130] error running task "Keypair/master" (9m48s remaining to succeed): error finding address for *openstacktasks.FloatingIP {"Name":"fip-api.sre.k8s.local","ID":"e1aeb0d6-217a-4119-a854-debce5135804","Server":null,"LB":{"ID":"f1c9a1e7-0bdc-4405-89ca-470e8e8bad73","Name":"api.sre.k8s.local","Subnet":"zone-1","VipSubnet":"9bff6415-ce08-493d-9b42-1b66120ef835","Lifecycle":"Sync","PortID":"08403111-d864-45d4-9ac4-f7cafe50b0eb"},"Lifecycle":"Sync"}: CreateFloatingIP: create floating IP failed: Resource not found
I0116 13:11:49.426025   72762 executor.go:103] Tasks: 82 done / 87 total; 5 can run
I0116 13:11:50.417529   72762 vfs_castore.go:729] Issuing new certificate: "master"
I0116 13:12:01.799328   72762 executor.go:103] Tasks: 87 done / 87 total; 0 can run
I0116 13:12:02.077788   72762 update_cluster.go:290] Exporting kubecfg for cluster

error getting ingress status: API Ingress Status not implemented for *openstack.openstackCloud
```

After that I can see that resources are installed, like instances networks etc. However, I do not see that anything is installed in machines? How I can get that working?

```
% openstack server list
+--------------------------------------+-------------------------------+--------+------------------------------------------+-------------------------+-----------+
| ID                                   | Name                          | Status | Networks                                 | Image                   | Flavor    |
+--------------------------------------+-------------------------------+--------+------------------------------------------+-------------------------+-----------+
| cca5a424-2c3c-4e4c-9f66-c1ea0e7c5baf | sre.k8s.local-bastions-1      | ACTIVE | sre.k8s.local=10.1.64.12, 10.222.134.126 | centos7-cloudimage-1809 | m1.medium |
| d6ee7498-31b5-443d-b463-a48f8fb5560d | sre.k8s.local-master-zone-2-1 | ACTIVE | sre.k8s.local=10.1.64.4                  | centos7-cloudimage-1809 | m1.medium |
| 4d7fa0e9-d2a4-4b3b-bd8d-d09344735698 | sre.k8s.local-master-zone-1-1 | ACTIVE | sre.k8s.local=10.1.64.11                 | centos7-cloudimage-1809 | m1.medium |
| f3b8e6ae-ca31-4c66-89b7-02c83f70ee70 | sre.k8s.local-nodes-1         | ACTIVE | sre.k8s.local=10.1.64.7                  | centos7-cloudimage-1809 | m1.medium |
| ca8bbd64-75ef-4ceb-b4da-f0a6bf5518f8 | sre.k8s.local-master-zone-3-1 | ACTIVE | sre.k8s.local=10.1.64.6                  | centos7-cloudimage-1809 | m1.medium |
+--------------------------------------+-------------------------------+--------+------------------------------------------+-------------------------+-----------+
```

**Comment by drekle** [comment]
> Example command returns
> 
> unknown flag: --etcd-storage-type
> 
> maybe something is not committed? :)

Yes, I had to touch create_cluster.go so I thought those changes would have been harder to get in when compared to these mostly openstack specific ones.

**Comment by drekle** [comment]
> Etcd (master) volume installation style does not work us. We have AvailabilityZone `nova` and then Volumetypes `zone-1` `zone-2` and `zone-3`

We'll have to figure out a mapping here.  I have the opposite problem as you, I have 3 compute AZ's with one Storage AZ.  This didn't make any sense to me either from a design standpoint and I am working to resolve this for us internally, however its totally valid in openstack.  We should think about how to create a mapping from compute to storage [to network]?

**Comment by drekle** [comment]
> Bastion host flavor?

I created a method for this in upup/pkg/fi/cloudup/openstack/utils.go `DefaultInstanceType`.  I may not be invoking it in this branch.

**Comment by drekle** [comment]
> ##### External network?
> At least we do have many networks in our openstack (for instance `ext-net` and `int-net`. How we can choose in which network this kops is exposed?

Mine is determined from the method GetExternalNetwork here:
upup/pkg/fi/cloudup/openstack/network.go

Let me know if this works for you

**Comment by drekle** [comment]
@zetaab If its easier I can create the PR with the full set of changes, and you'd be able to deploy.  Without etcd storage type as a command line argument you cannot provision the master volumes which allow etcd to be installed.  Without etcd the api server will just crash.  Additionally I have a branch for protokube which will set up GossipDNS between the nodes.

I suppose the most intrusive change outside of this branch is adding the argument to create_cluster.go.  I didn't want to create a large PR with changes to the other cloud flows.

**Comment by drekle** [comment]
> After that I can see that resources are installed, like instances networks etc. However, I do not see that anything is installed in machines? How I can get that working? Also notable thing is that all masters are installed in single zone, I had 3 zones when I created cluster.

Without etcd storage type as a command line argument you cannot provision the master volumes which allow etcd to be installed.  Without etcd the api server will just crash.  Additionally I have a branch for protokube which will set up GossipDNS between the nodes.

If you can ssh into the nodes I would imagine that you'd have an error with nodeup here:
```
journalctl -fu kops-configuration
```

File: upup/pkg/fi/cloudup/openstack/network.go [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+		err = networks.List(c.NetworkingClient(), networks.ListOpts{}).EachPage(func(page pagination.Page) (bool, error) {
+			var externalNetwork []NetworkWithExternalExt
+			err := networks.ExtractNetworksInto(page, &externalNetwork)
+			if err != nil {
+				return false, err
+			}
+			for _, externalNet := range externalNetwork {
+				if externalNet.External {
+					net = &externalNet.Network
+					return false, nil
```
Comment (drekle) [inline upup/pkg/fi/cloudup/openstack/network.go]: I am not sure why this works for me, however this should be `return true,nil`

**Comment by drekle** [comment]
> current code does not support multiple kops installations in same openstack project. This is because at least subnets are created without clustername prefix.

This is an issue I am okay with fixing, however I did not change any existing logic in `pkg/model/openstackmodel/network.go`

**Comment by zetaab** [comment]
@drekle I am using s3 kops state store. Works fine when creating cluster, everything seems fine but the instances are not just installed :)

**Comment by drekle** [comment]
> @drekle I am using s3 kops state store. Works fine when creating cluster, everything seems fine but the instances are not just installed :)

You might actually be getting further than I think you are.  I am curious if it makes it to the docker installation.  
*You might run the following on a master...*
```
journalctl -fu kops-configuration
journalctl -fu protokube
```

I doubt that you will get a full install without my protokube changes, as I require them for both DNS and etcd installation.  I will likely create another branch with my full set of changes thats rebased to master, as I want your delete changes as well anyway.

**Comment by zetaab** [comment]
@drekle can you close this PR, lets keep only one open

**Pull request closed without merging**